dotnet/src/Connectors/Connectors.Memory.Postgres/PostgresVectorStoreCollectionSqlBuilder.cs

```diff
@@ -0,0 +1,447 @@
+            }
+        }
+
+        if (keyProperty == null)
+        {
+            throw new ArgumentException("Record definition must have a key property.");
+        }
+
+        var keyName = keyProperty.StoragePropertyName ?? keyProperty.DataModelPropertyName;
+
+        StringBuilder createTableCommand = new();
+        createTableCommand.AppendLine($"CREATE TABLE {(ifNotExists ? "IF NOT EXISTS " : "")}{schema}.\"{tableName}\" (");
+
+        // Add the key column
+        var keyPgTypeInfo = PostgresVectorStoreRecordPropertyMapping.GetPostgresTypeName(keyProperty.PropertyType);
+        createTableCommand.AppendLine($"    \"{keyName}\" {keyPgTypeInfo.PgType} {(keyPgTypeInfo.IsNullable ? "" : "NOT NULL")},");
+
+        // Add the data columns
+        foreach (var dataProperty in dataProperties)
+        {
+            string columnName = dataProperty.StoragePropertyName ?? dataProperty.DataModelPropertyName;
+            var dataPgTypeInfo = PostgresVectorStoreRecordPropertyMapping.GetPostgresTypeName(dataProperty.PropertyType);
+            createTableCommand.AppendLine($"    \"{columnName}\" {dataPgTypeInfo.PgType} {(dataPgTypeInfo.IsNullable ? "" : "NOT NULL")},");
+        }
+
+        // Add the vector columns
+        foreach (var vectorProperty in vectorProperties)
+        {
+            string columnName = vectorProperty.StoragePropertyName ?? vectorProperty.DataModelPropertyName;
+            var vectorPgTypeInfo = PostgresVectorStoreRecordPropertyMapping.GetPgVectorTypeName(vectorProperty);
+            createTableCommand.AppendLine($"    \"{columnName}\" {vectorPgTypeInfo.PgType} {(vectorPgTypeInfo.IsNullable ? "" : "NOT NULL")},");
+        }
+
+        createTableCommand.AppendLine($"    PRIMARY KEY (\"{keyName}\")");
+
+        createTableCommand.AppendLine(");");
+
+        return new PostgresSqlCommandInfo(commandText: createTableCommand.ToString());
+    }
+
+    /// <inheritdoc />
+    public PostgresSqlCommandInfo BuildCreateVectorIndexCommand(string schema, string tableName, VectorStoreRecordVectorProperty vectorProperty)
+    {
+        var vectorColumnName = vectorProperty.StoragePropertyName ?? vectorProperty.DataModelPropertyName;
+        // Only support creating HNSW index creation through the connector.
+        var indexTypeName = vectorProperty.IndexKind switch
+        {
+            IndexKind.Hnsw => "hnsw",
+            _ => throw new NotSupportedException($"Index kind '{vectorProperty.IndexKind}' is not supported for table creation. If you need to create an index of this type, please do so manually. Only HNSW indexes are supported through the vector store.")
+        };
+
+        var indexOps = vectorProperty.DistanceFunction switch
+        {
+            DistanceFunction.CosineDistance => "vector_cosine_ops",
+            DistanceFunction.CosineSimilarity => "vector_cosine_ops",
```

Comment:
Cosine distance and similarity require different sorting on query.

Reply:
The way `BuildGetNearestMatchCommand` is coded, both cosine distance and cosine similarity use the same cosine distance query, which sorts in ASC order (which is required in pgvector to trigger the index; I'm not sure why, but ORDER BY distance DESC [doesn't hit the index](https://github.com/pgvector/pgvector?tab=readme-ov-file#why-isnt-a-query-using-an-index)). For cosine similarity, the cosine distance is wrapped in a subquery that transforms the distance to cosine similarity, but keeps the original ASC order for cosine distance, which will match a DESC ordering of the cosine similarity.